File: LeetCode/Python/203.remove-linked-list-elements.py
# @before-stub-for-debug-begin
from python3problem203 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=203 lang=python3
#
# [203] Remove Linked List Elements
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        dummyHead = ListNode(0)
        dummyHead.next = head
        tempHead = dummyHead

        while tempHead and tempHead.next:
            if tempHead.next.val == val:
                tempHead.next = tempHead.next.next
                continue
            
            tempHead = tempHead.next
        
        return dummyHead.next

# @lc code=end

File: LeetCode/CPP/203.remove-linked-list-elements.cpp
/*
 * @lc app=leetcode id=203 lang=cpp
 *
 * [203] Remove Linked List Elements
 */
#include <bits/stdc++.h>
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
using namespace std;
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeElements(ListNode *head, int val)
    {
        ListNode *dummy = new ListNode(-1, head), *prev = dummy, *tempHead = dummy->next;

        while (tempHead)
        {
            if (tempHead->val == val)
                prev->next = tempHead->next;
            else
                prev = tempHead;

            tempHead = tempHead->next;
        }
        return dummy->next;
    }
};
// @lc code=end